stl/inc/random

```diff
@@ -1116,27 +1118,48 @@ protected:
     _Post_satisfies_(this->_Idx == 0)
 
         void _Refill_lower() { // compute values for the lower half of the history array
-        size_t _Ix;
-        for (_Ix = 0; _Ix < _Nx - _Mx; ++_Ix) { // fill in lower region
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _Mx];
-        }
+        constexpr size_t _Wrap_bound_one = _Nx - _One_mod_n;
+        constexpr size_t _Wrap_bound_m   = _Nx - _M_mod_n;
 
-        for (; _Ix < _Nx - 1; ++_Ix) { // fill in upper region (avoids modulus operation)
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _Mx];
+        if constexpr (_M_mod_n == 0) {
+            for (size_t _Ix = 0; _Ix < _Wrap_bound_one; ++_Ix) { // fill in values
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+
+            if constexpr (_One_mod_n == 1) { // fill in _Ax[_Nx - 1]
+                constexpr size_t _Ix = _Wrap_bound_one;
+
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix - _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+        } else {
+            for (size_t _Ix = 0; _Ix < _Wrap_bound_m; ++_Ix) { // fill in lower region
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+
+            for (size_t _Ix = _Wrap_bound_m; _Ix < _Wrap_bound_one; ++_Ix) {
+                // fill in upper region (avoids modulus operation)
+                const _Ty _Tmp = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _M_mod_n];
+            }
+
+            if constexpr (_One_mod_n == 1) { // fill in _Ax[_Nx - 1]

```

Comment:
This says `// fill in _Ax[_Nx - 1]` but below you write `this->_Ax[_Ix]`, which is a bit confusing, same above

Reply:
Good catch, but I think this is OK - the commented conditions are when `_One_mod_n == 1`, so `_Ax[_Ix]` is always `_Ax[_Nx - 1]`. If I understand @statementreply's intent correctly, `_Ax[_Ix]` is being used to highlight the similar structure of these assignments. (For similar reasons, `_One_mod_n` is repeated even when it's known to be `1`.)